ISISComputingGroup/EPICS-ioc
# Ticket2111 instron stress rig

### Description of work

Instron stress rig EPICS IOC.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/2111

Ensure you also have the following changes when testing this:
https://github.com/ISISComputingGroup/ibex_gui/pull/505
https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/14
https://github.com/ISISComputingGroup/EPICS-instron/pull/1
https://github.com/ISISComputingGroup/EPICS/pull/83
https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/4

### Acceptance criteria

#### You will need to look at (at least) the following documents:
These are hosted on the sharepoint site under "Instron stress rig"
- "The spreadsheet" refers to `PVs.xlsx`
- "The reference manual" refers to `Stress Rig - GPIB.pdf`. Alternatively, there is a printed manual in the office.

#### Initialization
- [x] The initialization routine runs as documented on the spreadsheet. 

#### Reading from PVs
- [x] Reads from the device are done in a block
- [x] The watchdog is turned off before a block of reads (Interaction with device: `C904,0`)

#### Writing to PVs
- [x] For writes where it is needed, the device is put into computer control mode (`P909,1`) and taken out of computer control mode (`P909,0`) when it's finished

#### Rig commands
- [x] Stop button has PVs in place and sends a `C1,0` command to the rig. Note that the waveform PVs haven't been implemented yet so some of the records have TODOs on them. 
- [x] Panic stop behaves as expected.
- [x] Changing channels works as expected. (Caveat: channel commands are not yet in place, so cannot yet hide channels where `CHAN:TYPE:CHECK` is false).
- [x] Status bit 4 (Actuator active) is correctly parsed into `$(P)ACTUATOR:ON`
- [x] Status bits 9-12 are parsed into appropriate messages. See pages 5-42 and 5-43 of the reference manual. There is a mistake in the PDF version: `Power amplifier too hot` should report status code `0010` and not `1000`. This is consistent with both the labview driver and the printed manual.
- [x] You can set the type of movement using `$(P)MOVE:SP`. I've deviated slightly from the documents by combining several individual move commands into one PV with several possible values.
- [x] A utility PV called `$(P)MOVE:GO:SP` is provided, this stops and then starts movement.
- [x] If the rig reports a movement status of 2 or 5, then the movement is ended. Note that the movement status report from the rig is not the same as the movement type setpoint.
- [x] You can send arbitrary commands to the rig and get a response as expected

---

#### Code Review


### Functional Tests


### Final steps
